src/redisgears_python.c

```diff
@@ -2093,6 +2096,277 @@ static PyObject* createTensorFromValues(PyObject *cls, PyObject *args){
+    Py_INCREF(Py_None);
+    obj = Py_None;
+    
+clean_up:
+    RedisModule_CloseKey(key);
+    LockHandler_Release(rctx);
+    RedisModule_FreeString(rctx, keyNameStr);
+    RedisModule_FreeThreadSafeContext(rctx);
+    return obj;
+}
+
+static PyObject* msetTensorsInKeyspace(PyObject *cls, PyObject *args) {
+    verifyRedisAILoaded();
+    if(PyTuple_Size(args) != 1){
+        PyErr_SetString(GearsError, "Wrong number of arguments given to msetTensorInKeyspace");
+        return NULL;
+    }
+    // Input should be a Dictionary<string, tensor>.
+    PyObject* py_tensorDict = PyTuple_GetItem(args, 0);
+    if(!PyDict_Check(py_tensorDict)){
+        PyErr_SetString(GearsError, "Given argument is not of type PyDict");
+        return NULL;
+    }
+
+    PyObject* obj = NULL;
+    bool is_cluster = Cluster_IsClusterMode();
+    size_t len = PyDict_Size(py_tensorDict);
+    array_new_on_stack(RedisModuleKey*, STACK_BUFFER_SIZE, keys);
+    array_new_on_stack(RAI_Tensor*, STACK_BUFFER_SIZE, tensorList);
+    array_new_on_stack(const char*, STACK_BUFFER_SIZE, keyNameList);
+
+    Py_ssize_t i = 0;
+    PyObject *key, *value;
+    // Validate each key-value pair for <string, tensor> type.
+    while (PyDict_Next(py_tensorDict, &i, &key, &value)) {
+        if(!PyUnicode_Check(key)){
+            PyErr_SetString(GearsError, "key name argument must be a string");
+            goto clean_up;
+        }
+        if(!PyObject_IsInstance(value, (PyObject*)&PyTensorType)){
+            PyErr_SetString(GearsError, "Given argument is not of type PyTensor");
+            goto clean_up;         
+        }
+        tensorList = array_append(tensorList, ((PyTensor*)value)->t);
+        size_t keylen;
+        const char* keyNameCStr = PyUnicode_AsUTF8AndSize(key, &keylen);
+        if(is_cluster){
+            const char* keyNodeId = Cluster_GetNodeIdByKey(keyNameCStr);
+            if(!Cluster_IsMyId(keyNodeId)){
+                PyErr_SetString(GearsError, "Given key is not in the current shard");
+                goto clean_up;
+            }
+        }
+        keyNameList = array_append(keyNameList, keyNameCStr);
+    }
+   
+    RedisModuleType *ai_tensorType = RedisAI_TensorRedisType();
+    RedisModuleCtx* rctx = RedisModule_GetThreadSafeContext(NULL);
+    LockHandler_Acquire(rctx);
+    for(size_t i = 0; i < len; i++) {
+        const char *keyNameCStr = keyNameList[i];
+        RedisModuleString *keyNameStr = RedisModule_CreateString(rctx, keyNameCStr, strlen(keyNameCStr));
+        RedisModuleKey *key = RedisModule_OpenKey(rctx,keyNameStr, REDISMODULE_WRITE);
+        // Check if the key is empty or 
+        bool empty = RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY;
+        if(!empty && (RedisModule_ModuleTypeGetType(key) != ai_tensorType)) {
+            RedisModule_FreeString(rctx, keyNameStr);
+            PyErr_SetString(GearsError, "ERR Key is already set with non tensor values.");
+            goto clean_up;
```

Comment:
If you go to clean_up here do not free the rest of the keys in the keyNameList no? Isn't it better to use array_free_ex for keyNameList also?

Reply:
keyNemeList contains `const char*` I don't need to free it

Reply:
right. nice catch ..